# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/jsapi/PeerConnectionCtx.cpp
# Commit: 8d80a4ba914c
# Full Hash: 8d80a4ba914c6f81430ba1c55471a8d2b3aac60e
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 2
# Description:
#   Bug 1654112 - Create CallWorkerThread which implements AbstractThread for a TaskQueueWrapper. r=ng
#   
#   Ideally TaskQueueWrapper would implement both webrtc::TaskQueueBase and
#   AbstractThread, but webrtc::TaskQueueBase is not refcounted so implementing both
#   in the same class is not possible.
# ==============================================================================

diff -r e0a2e54e4de8 -r 8d80a4ba914c dom/media/webrtc/jsapi/PeerConnectionCtx.cpp
--- a/dom/media/webrtc/jsapi/PeerConnectionCtx.cpp	Thu Apr 08 11:31:43 2021 +0200
+++ b/dom/media/webrtc/jsapi/PeerConnectionCtx.cpp	Mon Apr 12 14:49:20 2021 +0200
@@ -12,6 +12,7 @@
 #include "common/browser_logging/WebRtcLog.h"
 #include "gmp-video-decode.h"  // GMP_API_VIDEO_DECODER
 #include "gmp-video-encode.h"  // GMP_API_VIDEO_ENCODER
+#include "libwebrtcglue/CallWorkerThread.h"
 #include "modules/audio_device/include/fake_audio_device.h"
 #include "modules/audio_processing/include/audio_processing.h"
 #include "modules/audio_processing/include/aec_dump.h"
@@ -168,16 +169,22 @@
 using namespace dom;
 
 SharedWebrtcState::SharedWebrtcState(
+    RefPtr<AbstractThread> aCallWorkerThread,
     webrtc::AudioState::Config&& aAudioStateConfig,
     RefPtr<webrtc::AudioDecoderFactory> aAudioDecoderFactory)
-    : mAudioStateConfig(std::move(aAudioStateConfig)),
+    : mCallWorkerThread(std::move(aCallWorkerThread)),
+      mAudioStateConfig(std::move(aAudioStateConfig)),
       mAudioDecoderFactory(std::move(aAudioDecoderFactory)) {}
 
+SharedWebrtcState::~SharedWebrtcState() = default;
+
 SharedModuleThread* SharedWebrtcState::GetModuleThread() {
+  MOZ_ASSERT(mCallWorkerThread->IsOnCurrentThread());
   if (!mModuleThread) {
     mModuleThread = webrtc::SharedModuleThread::Create(
         webrtc::ProcessThread::Create("libwebrtcModuleThread"),
         [this, self = RefPtr<SharedWebrtcState>(this)] {
+          MOZ_ASSERT(mCallWorkerThread->IsOnCurrentThread());
           mModuleThread = nullptr;
         });
   }
@@ -465,7 +472,15 @@
     audioStateConfig.audio_device_module =
         new rtc::RefCountedObject<FakeAudioDeviceModule>();
 
+    SharedThreadPoolWebRtcTaskQueueFactory taskQueueFactory;
+    auto callWorkerThread = WrapUnique(
+        taskQueueFactory
+            .CreateTaskQueueWrapper("CallWorker",
+                                    webrtc::TaskQueueFactory::Priority::NORMAL)
+            .release());
+
     mSharedWebrtcState = MakeAndAddRef<SharedWebrtcState>(
+        new CallWorkerThread(std::move(callWorkerThread)),
         std::move(audioStateConfig),
         already_AddRefed(CreateBuiltinAudioDecoderFactory().release()));
 